source/extensions/clusters/redis/redis_cluster_lb.cc

```diff
@@ -5,42 +5,99 @@ namespace Extensions {
 namespace Clusters {
 namespace Redis {
 
+bool ClusterSlot::operator==(const Envoy::Extensions::Clusters::Redis::ClusterSlot& rhs) const {
+  return start_ == rhs.start_ && end_ == rhs.end_ && master_ == rhs.master_ &&
+         slaves_ == rhs.slaves_;
+}
+
 // RedisClusterLoadBalancerFactory
-bool RedisClusterLoadBalancerFactory::onClusterSlotUpdate(
-    const std::vector<Envoy::Extensions::Clusters::Redis::ClusterSlot>& slots,
-    Envoy::Upstream::HostMap all_hosts) {
+bool RedisClusterLoadBalancerFactory::onClusterSlotUpdate(ClusterSlotsPtr&& slots,
+                                                          Envoy::Upstream::HostMap all_hosts) {
 
-  SlotArraySharedPtr current;
-  {
-    absl::ReaderMutexLock lock(&mutex_);
-    current = slot_array_;
+  // The slots is sorted, allowing for a quick comparison to make sure we need to update the slot
+  // array sort based on start and end to enable efficient comparison
+  sort(slots->begin(), slots->end(), [](const ClusterSlot& lhs, const ClusterSlot& rhs) -> bool {
```

Comment:
It's not particularly clear whether this sort() is worth it. Mostly I don't see any ordering dependencies in this code that the sort() might alleviate. Your thoughts, please?

Reply:
The sort() is useful for the current_cluster_slot_ && *current_cluster_slot_ == *slots.  

Let's say K(16384) slots are broken down to M ranges which are assigned to N shards(1 master + 0-n replicas).  M and N should normally be < 100 except for the largest clusters.

If we sort the M ranges based on the start and end for each range, then the comparison of whether allocation is changed is O(M log M) + O(M) without allocating additional space.  I can't be certain that "Cluster Slots" return a sorted range, as that's not part of the contract.  If it does return a sorted range, the sort function is going to be O(M) in most implementations, then it's a small cost to pay.